### ISSUE TITLE
NMatrix: Implement image processing methods

### ISSUE THREAD

**Issue by translunar** [SciRuby/nmatrix#199]
Here are two resources (courtesy [William Press and Numerical Recipes](http://www.nr.com/)).
- http://nr.com/CS395T/lectures2010/2010_18_WienerFiltering.pdf
- http://nr.com/CS395T/lectures2010/2010_19_LaplaceInterpolation.pdf

Those include a bunch of image processing methods which we'd like to see implemented in NMatrix, particularly fast Fourier transforms and Laplace interpolation.

Your mission, should you choose to accept it, is:
0. Open an issue in the tracker for the image processing method you select. You can also pick one that isn't listed above, but that you think might be useful (just make sure to explain why).
1. Write a blog entry, which we will post on sciruby.com, explaining how to go about performing the image processing task you select. Include pictures!
2. If you can't get it to work because NMatrix lacks some necessary functionality, submit that functionality as an issue. Try to get that functionality working and submit a pull-request.
3. Either email @mohawkjohn with your blog post or send a pull request in the sciruby.com repository. We'll edit it and post it with your name on it.

**Comment by eddiem3** [comment]
@andrewcsmith interesting, I'll check it out

**Comment by v0dro** [comment]
We might want to create this as another extension later...

**Comment by v0dro** [comment]
@mohawkjohn how about changing this issue and focus on making a plugin that will interface with fast image processing libraries?

**Comment by vaibhav-y** [comment]
I'd be interested in helping create such a plugin. (I'm comfortable with C if we need to implement the some of the functions ourselves)

Would an opencv wrapper be best for what we are looking at? Or would that be a bad idea due to the external dependency (a rather heavy one!)

**Comment by eddiem3** [comment]
@vaibhav-y I was working on that for a sec, but my harddrive fried before I pushed any code. I think it would be cool to do something especially with OpenCV.  Any thoughts @mohawkjohn ?

**Comment by vaibhav-y** [comment]
@eddiem3 I agree. OpenCV is really powerful.

I can setup a public workspace on c9.io and add you to that, if that's ok with you. That should be fail safe and also allows us to add more people for review / etc.

What I have in mind for the c-extension with a generic ruby wrapper (like a factory?) for whatever image processing library we choose

**Comment by mohawkjohn** [comment]
What kinds of methods from OpenCV would you be prioritizing? Can you give a few examples?

**Comment by eddiem3** [comment]
I'm thinking of starting with the basic image processing tools. Smothing,
Sobel Derivatives, Laplacian, Hough Line Transform, Canny Edge Detector

El lunes, 1 de febrero de 2016, Dr. John Woods <notifications@github.com>
escribió:

> What kinds of methods from OpenCV would you be prioritizing? Can you give
> a few examples?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SciRuby/nmatrix/issues/199#issuecomment-178014191>.
>


-- 
Eddie Massey III

Artificial Intelligence in Action Founder
Low Country Judicial Merit Selection Committee


803-290-2238
eddiemassey@ai-inaction.org
eddiemassey.com

**Comment by mohawkjohn** [comment]
Yeah. I think this would be a good candidate for plugins, and those sound like an excellent starting place.

**Comment by eddiem3** [comment]
So would a plugin be the same thing as a standalone gem?

**Comment by v0dro** [comment]
Go through the CONTRIBUTING file and see the links relating to making nmatrix plugins.

**Comment by eddiem3** [comment]
Ah got it thanks

**Comment by saru95** [comment]
I don't think a native - Ruby implementation of FFT and IFFT over a simple NMatrix currently exists !? It's really useful especially before applying filters. 
Currently, this is how it's done : 
1. Install FFTW.
2. Install Ruby bindings - fftw3.

I think this should be a feature. I'd like to know if a native FFT is being worked upon. Else I'd like to take this on ? :)

**Comment by v0dro** [comment]
There's also an nmatrix extension called nmatrix-fftw : https://rubygems.org/gems/nmatrix-fftw
